docs/developers/talawa-api/error-handling.md

```diff
@@ -0,0 +1,625 @@
+
+It resolves relations which exist in the business logic of the application with field resolvers only if the client specifically asks that relation to be resolved. This means lesser computation required to compute data not specifically needed by clients.(A resolver is basically a function with the same name as the field/relation that it resolves and returns)
+
+The data is returned in the form of a graph where every node is a meaningful object in context to the logic and the edges represent relations.
+
+In short GraphQl allows the clients to (These points will be important later on) -> 
+
+1. Query for a specific resolver which returns an object (node), meaningful to the api logic.
+2. Fetch for certain fields within said object(node) to optimize the request/response payload.
+3. In case of resolving relations(edges), making a nested query for interrelated objects(nodes).
+4. Each Node within the said nested query should be treated as individual entity in the graph.Error in one node should not affect other nodes in the graph. 
+
+This is perfect when everything runs as expected but what if it does not? How does GraphQl by default handle these 4 points for thinking in terms of graphs when an error is made?
+
+## Problem with default Graphql Errors
+
+We will understand more about default Graphql Errors with an example.
+
+Let's say I am making a query on the `signUp` mutation which accepts `email` , `name` and `password` and now we will look for the case where my password is shorter than 8 character, then the response returned by default ->
+
+```
+{
+  "data": {
+    "signUp": null
+  },
+  "errors": [
+    {
+      "path": [
+        "signUp"
+      ],
+      "locations": [
+        {
+          "line": 2,
+          "column": 3
+        }
+      ],
+      "message": "Password should be of atleast length 8"
+    }
+  ]
+}
+```
+
+
+Now the problems with this approach ->
+1. The `errors` key of a GraphQL response is not type safe. Graphql is loved because of the consistency and type safety it provides, we do not get the same with the `errors` array. They are specified, but introspection tells a client nothing about what might be found in them, they're harder to evolve, and harder to extend.
+2. By default if the `errors` array is present then the resulting query will return `null` in the `data` field. By that this means even if some nested resolver fails, that has the impact on the result of the root parent resolver (contradicts Introduction point 4).Essentially failure of a related node leads to the failure of parent node.This can potentially be a deal breaker if you're wanting to have errors returned as part of a mutation, but want to query for data on the result anyways. A common example is returning an error with some partial data for private Organization(name , location etc) even if a lot of background details like members , admins etc should not be returned.
```

Comment:
This is wrong. The nested resolver returning null only affects the parent resolver if the nested resolver's type in schema is defined as non-null. Since you've defined in schema that the nested resolver will never return null, when it does return null that's an error according to graphql engine. Since by the promise of schema it cannot return null for that nested resolver, the error traverses up to the parents until the engine finds a field which is defined as nullable in the schema.

By default the root `data` field is nullable in graphql, it's implemented in graphql engine core. So, null value for a nested resolver results in the whole data object being returned as null.

This situation can easily be avoided if you declare nested resolvers as nullable in schema.

```graphql
type User {
  posts: PostsConnection!
}
```

```graphql
type User {
  posts: PostsConnection
}
```

Notice the difference? The difference is `!` exclamation sign which makes it nullable/non-nullable. The concept of partial data is exactly this. If a nested resolver is nullable, you can return null or throw an error inside that resolver, and it will have no effect on the other fields. In case of throwing error the nested resolver field will have a value of null and in root errors array the error you threw in nested resolver will be present.

Reply:
@xoldyckk Oh, I had no idea that error could be avoided by making the return type nullable. 

Should I remove that sentence regarding nested resolver?

>By that this means even if some nested resolver fails, that has the impact on the result of the root parent resolver (contradicts Introduction point 4).Essentially failure of a related node leads to the failure of parent node.This can potentially be a deal breaker if you're wanting to have errors returned as part of a mutation, but want to query for data on the result anyways. A common example is returning an error with some partial data for private Organization(name , location etc) even if a lot of background details like members , admins etc should not be returned.

Reply:
@kb-0311 If you had no idea about it then you should learn more about it and it's workings. Might change the mental model of how you're currently perceiving things, will probably let you think of more flexible approaches.

Reply:
@xoldyckk Yeah that is the tricky part there is not much info or a universally agreed upon approach regarding Graphql Error handling other than the stage 6a one and stage 5 one and a lot of stuff regarding best practises for error handling keeps changing. 

Even if we consider the nullable point you raised  I doubt it would be enough to completely change the approach ->
>In case of throwing error the nested resolver field will have a value of null and in root errors array the error you threw in nested resolver will be present. The error will also exactly mention which field the error occured in, in its path field. 

There is the matter of Apollo client since the default graphql errors array in there specifies the path and considers that as the differentiating factor , what happens when we are restructuring `data` and `error` object from a `useQuey` hook?  

The client apps will show an error for the whole query since any error in the root error array will cause it to fail.

There are ways around this on the client apps ofc but I cannot think one which would be simpler than this one currently.